File: CodeForces/Contest/Dytechlab Cup 2022 (Div 2 + 1)/DDTb.cpp
#include <bits/stdc++.h>

using namespace std;

struct Edge{
   int a;
   int b;
   long long w;
};

int main(){
   ios::sync_with_stdio(false);
   cin.tie(0);
   
   int t;
   cin >> t;
   for(int q = 0; q < t; q++){
      int n,m;
      cin >> n >> m;
      
      //cout << n << " " << m;
      
      vector<vector<long long>> d(n+1,vector<long long>(n+1,1000000000000000LL));
      
      vector<Edge> edges;
      
      for(int k = 0; k < m; k++){
         int a,b;
         long long w;
         cin >> a >> b >> w;
         //cout << "edge: " << a << " " << b << " " << w << endl;
         Edge e;
         e.a = a;
         e.b = b;
         e.w = w;
         edges.push_back(e);
         
         d[a][b] = min(d[a][b],w);
         d[b][a] = min(d[b][a],w);
      }
      
      for(int k = 1; k <= n; k++) d[k][k] = 0LL;
      
      for (int k = 1; k <= n; ++k) {
          for (int i = 1; i <= n; ++i) {
              for (int j = 1; j <= n; ++j) {
                  d[i][j] = min(d[i][j], d[i][k] + d[k][j]); 
              }
          }
      }
      
      for(int k = 1; k <= n; k++){
         for(int j = 1; j <= n; j++){
            cout << d[k][j] << " ";
         }
         cout << endl;
      }
      
      
      long long answer = LLONG_MAX;
      for(int k = 0; k < m; k++){
         //cout << edges[k].a << " " << edges[k].b << " " << edges[k].w << endl;
         long long move = (long long)min(d[1][edges[k].a]+d[n][edges[k].b],d[1][edges[k].b]+d[n][edges[k].a]);
         answer = min(answer,move*edges[k].w);
      }
      
      cout << answer << "\n";
      
       
      
   }
   
   
   return 0;
}

File: CodeForces/Contest/Dytechlab Cup 2022 (Div 2 + 1)/DDTb.java
//make sure to make new file!
import java.io.*;
import java.util.*;

public class DDTb{
   
   
   
   public static void main(String[] args)throws IOException{
      BufferedReader f = new BufferedReader(new InputStreamReader(System.in));
      PrintWriter out = new PrintWriter(System.out);
      
      int t = Integer.parseInt(f.readLine());
      
      for(int q = 1; q <= t; q++){
      
         StringTokenizer st = new StringTokenizer(f.readLine());
      
         int n = Integer.parseInt(st.nextToken());
         int m = Integer.parseInt(st.nextToken());
         
         long[][] d = new long[n+1][n+1];
         
         for(int k = 1; k <= n; k++){
            for(int j = 1; j <= n; j++){
               d[k][j] = 1000000000000L;
            }
            d[k][k] = 0L;
         }
         
         Edge[] edges = new Edge[m];
         
         for(int k = 0; k < m; k++){
            st = new StringTokenizer(f.readLine());
         
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            long w = Long.parseLong(st.nextToken());
            
            edges[k] = new Edge(a,b,w);
            
            d[a][b] = Math.min(d[a][b],1);
            d[b][a] = Math.min(d[b][a],1);
         }
         
         
         for (int k = 1; k <= n; ++k) {
            for (int i = 1; i <= n; ++i) {
               for (int j = 1; j <= n; ++j) {
                  d[i][j] = Math.min(d[i][j], d[i][k] + d[k][j]); 
               }
            }
         }
      
         long answer = Long.MAX_VALUE;
         for(int k = 0; k < m; k++){
            //cout << edges[k].a << " " << edges[k].b << " " << edges[k].w << endl;
            long move = (long)Math.min(d[1][edges[k].a]+d[n][edges[k].b],d[1][edges[k].b]+d[n][edges[k].a])+1;
            out.println(edges[k].a + " " + edges[k].b + " " + move);
            answer = Math.min(answer,move*edges[k].w);
         }
      
         out.println(answer);
      
      
      }
      
      
      
      
      out.close();
   }
   
   public static class Edge{
      int a;
      int b;
      long w;
      public Edge(int c, int d, long e){
         a = c;
         b = d;
         w = e;
      }
   }
   
      
}